### DISCUSSION TITLE
Per-item inputs/data on the Cart block

### DISCUSSION THREAD

**Discussion by opr** [woocommerce/woocommerce#45032] [WooCommerce, Blocks, and Site Editing]
Each item in the cart may need controls to change settings, or to display some data. For example a checkbox to indicate a product is a gift, or a text input to enter a custom note for the gift.

![image](https://github.com/woocommerce/woocommerce-blocks/assets/5656702/d0b97de6-9c37-46ab-bf5f-0dad420f0b02)

We plan to look into how we can achieve this and this GitHub discussion will serve as the main discussion point for this extensibility interface.

**Requirements**

- Developers must be able to add one or more controls to specific cart items to control a value.
- The configuration requirements must be available to consumers of the Store API, e.g. they must know that there should be a checkbox rendered for each cart item with the label “This is a gift”
- An extension must be able to conditionally show or hide an individual field (this may depend on a value of one of its other fields too).

**Comment by helgatheviking** [comment]
I also need the ability to add a button that will load further fields. Specifically my free gift coupons plugin can add a variable product and then give the customer the attributes form in the cart so they can pick which attribute. Example: free gift T-shirt. Customer gets to pick size.

**Reply by opr** [reply]
Thanks for adding this use case. This seems like it should be doable, but we would need to confirm this. Ultimately the button would act as a toggle for showing the additional fields. Does the button disappear after the shopper has clicked it, or can they click it again to collapse the attribute field?

Why is the button necessary by the way, when the free gift t-shirt is added why not just show the fields immediately? (Just trying to understand the shopper flow here)

**Reply by helgatheviking** [reply]
Yes, when the coupon is added and the gift product is added then the fields should show automatically. But you can have a button there in case the user needs to edit their choices later. 

Currently in it cart block, the button links back to the single product page to make selections there.

**Reply by opr** [reply]
OK, thanks for the additional context.

Since product variations are treated as a different product we will need to consider the specific use case of changing variations from the cart. This could mean a new endpoint where the old item is removed and the new item is added.

A concern we need to be careful of is things like coupons only being valid on specific variations, quantity limits, and purchase limits. We should also be careful about stock levels, e.g. if the T-shirt is not available in XL, do we show that option anyway and just disable it?

**Reply by helgatheviking** [reply]
For an overview of how it works currently visit:
https://dev-mix-and-match.pantheonsite.io/shop/
Then add something to the cart and visit the cart page. Then apply the coupon `variablegift`. A variable product will be added to the cart and the add to cart form should automatically display so you can select your variation. 

As for certain variations being unavailable, the variable add to cart form should handle that... So I'm relying on that to prevent gifting an out of stock variation.

**Reply by helgatheviking** [reply]
Adding a recording of current/classic functionality https://share.zight.com/GGuBqdDR

**Reply by pierorocca** [reply]
The recording is super helpful thank you Kathy. Curious if a modal could work in this context to show the user the variable product? There's a lot of screen shifting in the example and I mean this in the kindest possible way, it lacks a modern feel to the UX. A modal prompts the user that they have a task to complete, does it in a way that's distraction free, and keeps the existing cart functionality intact which helps with interoperability between plugins. It would be a better mobile experience and having the full variable product accessible would allow the user to see images of the variations on a carousel rather than being limited to a single, small image on the cart.

Thoughts on that option? @opr is something like that even technically feasible?

**Reply by opr** [reply]
@pierorocca I like the idea of a modal. Kathy's example does a fine job of showing the use case, but we can certainly refine the UX to take up less space and, as you say, keep the shopper focused. On the topic of whether it's technically feasible, if it's specifically the edit variant functionality, then I'm like 99% sure it is, but we would need a tech spike to 100% confirm this.

**Reply by pierorocca** [reply]
Yes the example did a terrific job of articulating the need. If I understood it correctly, the coupon makes the shopper eligible for a specific variable product. My idea here is we show that variable product page, or a variant of it, in the modal so the user has full access to the product images, description, reviews/ratings, etc. A simple product obviously is a lot more straightforward, especially if it's a promotional item. e.g. free tote bag.

The edit options button - This is a good consideration first for the cart block and should/could variable products be edited from the cart? Some sites allow it. Others force shoppers to delete the item and re-add it with the updated options. If we support it natively, then extensibility is not relevant. If we continue to not support it, do we open up extensibility and can it be achieved without sacrificing interoperability would be the types of questions I'd be looking to answer. "Extensibility that doesn't sacrifice interoperability" is my mantra for 2025.

**Reply by helgatheviking** [reply]
Yes, the video was largely to show the need for placing a button on the per-item basis. Currently, my approach is pretty janky as there's no built-in way to do this for the cart block.

I like the idea of a modal though! I've been attempting to force a react portal to keep parity, but the modal does sound like better UX. I will have to look in to that when I come back to this project. 

Underneath I had been removing the product and re-adding the product to the cart with the newly selected attributes.. which should still be doable via the Store API.

**Comment by marcusig** [comment]
I have a similar situation, where I need to be able to add a link to a cart item (previously discussed (https://github.com/woocommerce/woocommerce-blocks/issues/10435)[in this issue] which was apparently not brought over to the mono-repo).

Cart Shortcode: | Cart Block:
-- | --
![image](https://github.com/woocommerce/woocommerce/assets/11990914/ff6efca5-9981-4589-b047-56196f50ea48) |![image](https://github.com/woocommerce/woocommerce/assets/11990914/3dc4de15-6814-4026-921b-fa311ce2cd7c)

I won't go over all the differences as I have opened separate issues for them. 

But In the shortcode, there is an "Edit configuration" link, which is currently added at the end of the item meta.
I could also add it next to the "Remove item" button, if there was such a possibility.

**Reply by opr** [reply]
Hi @marcusig this is a great example of a use-case, thanks for sharing! We will keep it in mind when it comes to developing this feature.

**Reply by helgatheviking** [reply]
Similar to Marcus, I have an "edit" link needed in Name Your Price and "Mix and Match" though neither launch an edit "in place" in the cart and rather, redirect the user back to the single product page.

**Comment by pinalj** [comment]
Usecase: 
We're looking to add a 'Edit Delivery Date' button for each item in the Cart Blocks. Something like as shown [here](https://prnt.sc/_fDyMTp4jT4G)

When clicked it will display a popup allowing the user to edit the delivery date.

**Reply by opr** [reply]
Thank you for sharing your use case! This is a great example. We will keep it in mind when developing this feature 🙇🏼

**Comment by a3rev** [comment]
Hello @opr 

Is there any update on work been done on this? You opened this ticket 6 months ago, and since that time there have been 6 major version releases 8.3 to 8.8 but nothing  to facilitate Per-item inputs/data on the Cart block. This was the issue that I opened 5 weeks ago [#45537](https://github.com/woocommerce/woocommerce/issues/45537) and you closed and pointed me here. But there does not seem to be any activity at all on this issue with the last update posted before I post issue #45537.

From a developers point of view it does not seem an unreasonable expectation that we should be able to filter on the Cart and Checkout blocks to the same extent that we can on the legacy templates. I mean I would NOT have thought this was an 'edge case' request? 

Are there big issues with implementing this?

Steve

**Comment by nielslange** [comment]
Hello @a3rev,

I understand your frustration regarding the lack of updates on per-item inputs/data for the Cart block. I wanted to share some insights from a recent initiative that might shed light on the complexities involved in implementing such features.

During a recent meetup, @opr and I were looking into this issue. It turned out that it's not trivial to implement this option. We were facing various issues such as managing the DOM structure changes without breaking layout and styling, and the complexities of passing unique props to cloned blocks for each cart item on the frontend.

We are continuing to explore different approaches, including the potential use of block hooks, before considering alternatives like Slot/Fill. Unfortunately, at this time, we cannot provide a specific timeline for when these features will be implemented.

That said, your input and use case are invaluable to us, and we are committed to finding a viable solution that allows developers like yourself more flexibility in customizing the WooCommerce Cart and Checkout blocks.

**Reply by a3rev** [reply]
Hello @nielslange 

Thank you for your response and the update. Good to know that it is being worked on. 
Steve

**Comment by akansh-sirohi** [comment]
Use Case:
We are endeavoring to develop a plugin that supports custom linked products. Our aim is to display these products as a unified entity, rather than as two separate items. Therefore, it would be beneficial if there were a method to filter the specific products we wish to display and determine how they will be rendered on the page. Any assistance in achieving this would be greatly appreciated.

**Comment by Ninodevo** [comment]
Hi @opr! Adding an additional use case for your reference:

We would like to add product add-ons to cart items. The user would see an action link/button on the cart item to add or remove add-ons related to the item. This action would add or remove items in the cart or modify the current item in the cart.

We currently have this implemented with a custom block below the cart items, but we would like to have it per cart item.

**Comment by paulg-rymera** [comment]
We are also looking for the ability to customize the quantity input step to whatever backend setting conditions the site store might have.

**Reply by opr** [reply]
@paulg-rymera thanks! Can you give some examples of what settings the store may have? We do have hooks available that control the quantity inputs: https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/docs/third-party-developers/extensibility/hooks/filters.md#woocommerce_store_api_product_quantity_value_type

**Reply by paulg-rymera** [reply]
In our plugin, we are providing users a way to customize the quantity step for specific wholesale products. So. for example, Product A might require a minimum quantity of `10` and _should_ be in increments of `5` before wholesale pricing kicks in. Then Product B, might have a minimum requirement of `20` and increments of `10`.

We have a way already with the minimum requirements, but we are looking for a way on how to make it work with the quantity step requirements for the cart block (this works on legacy).

**Comment by marcusig** [comment]
Hi @opr, 
Just checking in to see if there has there been any progress towards a solution for this issue?

**Reply by opr** [reply]
Hi @marcusig thanks for asking. Sorry but there hasn't been any progress on this yet, the issue will be updated when we start working on it so feel free to subscribe for updates. It's a pretty hard problem to solve right now and we've been prioritising other features.

**Reply by marcusig** [reply]
Ok, thanks for the update :)

**Reply by marcusig** [reply]
Hi @opr 
Sorry to ping again about this :)
I don't know if there's been any progress with this issue, but I read [this Slack thread](https://woocommercecommunity.slack.com/archives/C03MCL9RTAR/p1736331425790039?thread_ts=1735916187.520329&cid=C03MCL9RTAR) where @senadir shares a solution to manipulate the cart items. 

However it feels a bit hacky and unnecessarily complex . 

Would it not be possible to have a simple hook like this?

```JS
wp.hooks.doAction( 'cart_item_rendered', dom_element, cart_item_key, cart_item_data );
```

**Reply by opr** [reply]
Hi @marcusig there are some tough problems to get around before we can open the cart items up for extensibility, you're right, the idea Nadir shared in Slack (posting here for posterity, since Slack messages get deleted after 90 days) is a little hacky, but it's the best method for now.

> Nadir Seghir, Woo Engineer
I will try to get you an example soon, but the gist of it is:
Use the JS calssname filter to add cart item key as a classname.
Use said classname to fetch the corresponding dom element of each item.
Using cart store data, and the dom just insert whatever UI with whatever data you have, using things like appendChild, 
rest of magic is around knowing when to run stuff, how to react to things, mostly via using wp.data.subscribe and whatnot
custom data you need for each item can be passed from store api via extendSchema like `woocommerce_store_api_register_endpoint_data` on cart item

If we did implement something like you are asking for that was just a proxy to Nadir's idea, then we would be tied into supporting that forever, which is not something we can commit to at the moment. At least not until the future around how WooCommerce blocks uses the interactivity API and block hooks becomes clearer.

Sorry this isn't a very helpful answer but hopefully you understand a bit more about our limitations to progressing this further.

**Reply by helgatheviking** [reply]
> there are some tough problems to get around before we can open the cart items up for extensibility, 

Could to elaborate on what the blocking issues are? 

I never did attempt the solution proposed in slack... Could someone pull together a brief example?

**Reply by marcusig** [reply]
I'm trying to use this method, however the subscribed event is triggered before the classname filter. So do we have to use a MutationObserver to make the required DOM changes, or is there an other event or hook we can use?

My JS: 

```JS
(function() {
	// document.addEventListener( 'DOMContentLoaded', function() { 
	if ( 'undefined' == typeof window.wc.blocksCheckout ) return;

	const { registerCheckoutFilters } = window.wc.blocksCheckout;

	if ( 'undefined' == typeof registerCheckoutFilters ) return;

	registerCheckoutFilters( 'my-plugin', {
		// Adjust cart item class of the cart line items.
		cartItemClass: ( value, extensions, args ) => {
			console.log( 'registerCheckoutFilters - cartItemClass - add cartItem.key as class' );
			
			value += ' item-' + args.cartItem.key;

			return value;
		}	
	} );


	const { select, subscribe } = window.wp.data;
	const cartStoreKey = window.wc.wcBlocksData.CART_STORE_KEY;
	const unsub = subscribe( onCartChange, cartStoreKey );

	function onCartChange() {
		const cart = select( cartStoreKey ).getCartData();
		console.log( '¡ cart change !', cart.items.length );
		cart.items.forEach( item => {
			console.log( 'cart item', item.key, document.querySelector( '.item-' + item.key ) );
		} );
	}
} ) ();
```

Console Logs:
```
¡ cart change ! 1
cart item 40ac0b1a075a22ed16cd98b829e68546 null
registerCheckoutFilters - cartItemClass - add cartItem.key as class
```

**Reply by opr** [reply]
> I never did attempt the solution proposed in slack... Could someone pull together a brief example?

@helgatheviking I think that is based on the way you implemented it, I recall you are rendering a React portal in the cart items, am I remembering wrong?

> Could to elaborate on what the blocking issues are?

We did spend a chunk of time looking into this at a team meetup last year, and we were blocked by issues with the editor and how we render the preview. Ideally merchants would be able to add inner blocks to their cart line items, but since the preview renders two line items, it wasn't possible to have an area where inner blocks can be added _and_ make them show up in both line items in the preview. I think just showing them in one would be a bad UX for the merchant. The ideal solution would be for block hooks to work here, but they don't, I think this was due to the way our blocks render actual React components on the front end.

In terms of actually allowing custom content to be rendered, it's doable, we could use slot fills but this is not an ideal solution as it only works for extensions/developers, and wouldn't translate to having an editable inner blocks area in the editor. That would essentially mean us supporting two systems going forward.

With all this said, I do have a meeting next week to discuss extensibility some more and I will bring this up as one of the biggest pain points merchants and developers bring up to us. cc @pierorocca

**Reply by helgatheviking** [reply]
Thanks for the explanation!

I was actually filtering the rest api short description to add a button... i think @senadir's proposal was an alternative to that because it is admittedly very hacky. the button launches the portal which I did manage and at some point i will pick that project back up. but like Marc, I was trying to add content on a per-cart-item basis.

**Reply by senadir** [reply]
> it's doable, we could use slot fills

I think Slotfills are not possible actually in that use case (loop)

**Reply by helgatheviking** [reply]
>I think Slotfills are not possible actually in that use case (loop)

I also was just dabbling with SlotFills and loops and couldn't make it work... I would agree it's not possible.

**Reply by marcusig** [reply]
@opr 
>  it only works for extensions/developers, and wouldn't translate to having an editable inner blocks area in the editor. That would essentially mean us supporting two systems going forward.

Is your current goal to give the possibility to edit the cart item layout in the editor?
While I see the benefit of that, In my mind, because the cart line item depends on the product data, product type and whatever product extensions exist, there is an (more?) important need for developer extensibility. 

But again, I don't have all the user feedback that you have, so I could be wrong.

**Reply by pierorocca** [reply]
Gifting is a common use case so I propose this is something we need to look at first as a core feature rather than as an extensibility feature. Otherwise, we can end up with x number of interpretations with different design language and styling that clashes with the design system. IIRC gifting and gift messaging has been a core gap for a long time and is one of the top uses of the Checkout Field Editor.

The use cases I've identified so far in this thread are:

**Use Cases**
- Gifting and gift messaging - Yes an extremely common use case. Strong case for this to be a core supported feature.
- Add variable product when coupon code is applied - Less common. 
    + Editing in the cart is not as desirable of a UX as popping up the variable product page in a modal exposing the full capability like image carousel, reviews, description, dimensions, etc..
- Editing variable product options from the cart - Yes common use case. How might we support this and make it configurable? A modal may work here. This would need design exploration.
- Quantity stepper increments that are configurable - I see that the cart supports the min/max plugin that allows for grouped quantities. e.g. 5. What's the technical barrier here that is preventing an extension from doing the same? 

Are there other use cases I've missed that support the need for this particular extensibility?

**Reply by opr** [reply]
> Gifting and gift messaging - Yes an extremely common use case. Strong case for this to be a core supported feature.

Yep, agreed we should build this in.

> Add variable product when coupon code is applied - Less common.

This _should_ already be possible, when the coupon is applied, a product can be added to the cart on the server. When the server responds with the updated cart, the client-side cart should be updated to show the new item.

> Editing in the cart is not as desirable of a UX as popping up the variable product page in a modal exposing the full capability like image carousel, reviews, description, dimensions, etc..

Yeah, I love the modal idea, I think a generic solution to allow a modal containing any content to show when a particular trigger is clicked would be nice too, but definitely out of scope for a first iteration targeting variable products.

> Quantity stepper increments that are configurable - I see that the cart supports the min/max plugin that allows for grouped quantities. e.g. 5. What's the technical barrier here that is preventing an extension from doing the same?

There is no barrier to this, any plugin can modify the `woocommerce_store_api_product_quantity_[maximum/minimum/multiple_of]` filter to change this behaviour.

**Reply by pierorocca** [reply]
> There is no barrier to this, any plugin can modify the woocommerce_store_api_product_quantity_[maximum/minimum/multiple_of] filter to change this behaviour.

Perfect thanks @opr. The product quantity is the right place for that change, so as to avoid introducing a brand new control that replicates (and proliferates/fragments) what's already possible.

**Comment by LuigiPulcini** [comment]
Has there been any progress or updates on this issue? As of June 2025, cart items metadata in the cart block escapes the HTML markup content. Due to the lack of HTML support, we are forced to recommend that our customers use the Classic Cart block or the `[woocommerce_shortcode]` directly.

I can add our use case, in case it can reinforce all the other use cases.
One of our plugins allows adding several custom options to a product and generates a customized version of the product image. The images uploaded by shoppers, together with the preview image generated by our plugin are displayed inside the cart item so that shoppers can confirm their configuration and shop owners cn receive the correct data in the order.

Currently, none of the features available in the classic cart are supported in the cart block, which renders it completely unusable for our customers.

![image](https://github.com/user-attachments/assets/3fda9144-d44b-412b-8cda-1170fe6a313e)

**Comment by rjm87** [comment]
Curious on any possible updates on this?

**Comment by ralucaStan** [comment]
In https://github.com/woocommerce/woocommerce/pull/61437 we've allowed some HTML elements in the `item_data` metadata: `'a', 'b', 'em', 'i', 'strong', 'br', 'abbr', 'span'`.

For now, we're keeping this as a list of allowed elements. 
If you were using `item_data` to add other HTML elements please share your use case, if you haven't already.

**Comment by helgatheviking** [comment]
Hi all... I am back at digging in to this and have found some success with a frankenstein approach:

https://www.loom.com/share/6d843619f6f14cc8adcd73f906c21cdc

Lemme try to outline my approach:

1. I have created an iAPI block
2. Technically it is also registered as a cart block integration via `IntegrationInterface` though I am not certain this is required.
3. I use `render_block_woocommerce/cart` filter to add my new block to the end of the cart:

```
add_filter( 'render_block_woocommerce/cart', function ( $block_content ) {
		// Manually call wp_interactivity_process_directives() to process the directives.
		// Normally, we would tell the woocommerce/cart block to support interactivity (via add_custom_metadata method), but we can't do that since the woocommerce/cart block has no [`block.json` Waiting on my PR for that too](https://github.com/woocommerce/woocommerce/pull/56142) :).
		$block_content .= wp_interactivity_process_directives( do_blocks('<!-- wp:my-plugin/my-block /-->') ); 
	}

	return $block_content;
} );
```

Then in my block's `view.js` module, I use a mutation observer to move it into the correct location. Now this does rely on this pending [PR](https://github.com/woocommerce/woocommerce/pull/63090) to give the cart line item rows a unique data attribute. I could do this via the cart class filter if absolutely necessary. 

Locally, I am have also modified the[ `cart-update-item` store endpoint](https://github.com/woocommerce/woocommerce/blob/c2eb100a2884950a5d6339e5411b6534920c34ac/plugins/woocommerce/src/StoreApi/Routes/V1/CartUpdateItem.php#L70-L78) as currently it only modifies the quantity. I'm still thinking over the best way to do this, but will send a PR. I could also create a standalone rest route, but it does seem like `cart-update-item` should be able to do more. 

Then in my block I make a call to the update endpoint:

```
// TODO: Make API call to update cart item.
const response = yield fetch(`${ restUrl }wc/store/v1/cart/update-item`, {
	method: 'POST',
	headers: {
		'Content-Type': 'application/json',
		Nonce: nonce,
	},
	body: JSON.stringify({
		key: item.key,
		myInput: inputValue
	}),
});
```

and if the response is OK then I trigger the refresh.
```
wp.data.dispatch('wc/store/cart').invalidateResolutionForStore();
```

It _works_ but it's a little frankenstein-ish in my opinion, depends on multiple PRs including one I havent sent yet.. and... it's slow because it has to hit the update endpoint and then fetch the refreshed cart. There's gotta be a better way to do that, but for now the point was to insert my content so it's just a perk that it actually does something.

I'm keep to take a look at the mini cart and see what lessons can be learned from that since it is already using the iAPI.

**Comment by helgatheviking** [comment]
I have sent a [PR](https://github.com/woocommerce/woocommerce/pull/63296) with what I think is a better approach as it doesn't use frankencode :sweat_smile:
